src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,193 @@
+            Node<T> previous = current.prev;
+
+            newNode.next = current;
+            newNode.prev = previous;
+
+            if (previous != null) {
+                previous.next = newNode;
+            }
+            current.prev = newNode;
+        }
+
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null || list.isEmpty()) {
+            return; // Якщо список порожній або null, нічого не додаємо
+        }
+
+        for (T element : list) {
+            add(element); // Додаємо кожен елемент у кінець поточного списку
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index); // Перевірка валідності індексу
+
+        Node<T> current = head; // Починаємо з голови списку
+        for (int i = 0; i < index; i++) {
+            current = current.next; // Переходимо до наступного вузла
+        }
+
+        return current.value; // Повертаємо значення вузла за індексом
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateIndex(index); // Перевіряємо, чи індекс валідний
+
+        Node<T> current = head; // Починаємо з голови списку
+        for (int i = 0; i < index; i++) {
+            current = current.next; // Переходимо до потрібного вузла
+        }
+
+        T oldValue = current.value; // Зберігаємо попереднє значення
+        current.value = value; // Замінюємо значення на нове
+        return oldValue; // Повертаємо попереднє значення
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);// Перевірка валідності індексу
+
+        if (index == 0) {
+            // Якщо видаляємо перший елемент
+            T removedValue = head.value; // Зберігаємо значення вузла
+            head = head.next; // Змінюємо посилання на наступний вузол
+            size--; // Зменшуємо розмір списку
+            return removedValue; // Повертаємо видалене значення
+        }
+
+        Node<T> current = head;
+        for (int i = 0; i < index - 1; i++) {
+            current = current.next; // Переходимо до вузла перед тим, який потрібно видалити
+        }
+
+        T removedValue = current.next.value; // Зберігаємо значення вузла, який видаляємо
+        current.next = current.next.next; // Пропускаємо вузол, який потрібно видалити
+        size--; // Зменшуємо розмір списку
+        return removedValue; // Повертаємо видалене значення
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        if (head == null) {
+            return false; // Якщо список порожній, повертаємо false
+        }
+
+        // Якщо потрібно видалити перший елемент
+        if (head.value == null ? object == null : head.value.equals(object)) { // Обробка null
+            head = head.next; // Змінюємо посилання на наступний вузол
```

Comment:
Similar to the `remove(int index)` method, when removing the first element, if the list becomes empty, you should also set `tail` to `null`.